src/Swift.Bindings/src/Emitter/StringEmitter/Handler/TypeHandler.cs

```diff
@@ -588,28 +505,46 @@ private void EmitPrivateConstructor()
         /// </summary>
         private void WriteMarshalToSwiftFrozenStruct()
         {
-            string payloadName = "this";
             TypeRecord typeRecord = _typeDatabase.GetTypeRecordOrThrow(_structDecl.SwiftTypeName);
             if (MarshallingHelpers.IsFrozenStructProjectedAsClass(typeRecord))
             {
-                payloadName = "_payload";
-            }
+                var text = $$"""
+                unsafe void ISwiftObject.MarshalToSwift(Span<byte> swiftDestSpan)
+                {
+                    var metadata = SwiftObjectHelper<{{_structDecl.Name}}>.GetTypeMetadata();
+                    Debug.Assert((int)metadata.Size == swiftDestSpan.Length, $"Span size does not match type size, Expected: {(int)metadata.Size}, Actual: {swiftDestSpan.Length}");
+                    fixed (void* swiftDest = swiftDestSpan)
+                    {
+                        // Ensure that the instance is valid before making copy
+                        bool success = false;
+                        _payload.DangerousAddRef(ref success);
+                        metadata.ValueWitnessTable->InitializeWithCopy((void *)swiftDest, _payload, metadata);
+                        if (success)
+                            _payload.DangerousRelease();
+                    }
+                }
+                """;
 
-            var text = $$"""
-            IntPtr ISwiftObject.MarshalToSwift(IntPtr swiftDest)
+                _writer.WriteLines(text);
+            }
+            else
             {
-                var metadata = SwiftObjectHelper<{{_structDecl.Name}}>.GetTypeMetadata();
-                unsafe {
-                    fixed (void* payload = &{{payloadName}})
+                var text = $$"""
+                unsafe void ISwiftObject.MarshalToSwift(Span<byte> swiftDestSpan)
+                {
+                    var metadata = SwiftObjectHelper<{{_structDecl.Name}}>.GetTypeMetadata();
+                    Debug.Assert((int)metadata.Size == swiftDestSpan.Length, $"Span size does not match type size, Expected: {(int)metadata.Size}, Actual: {swiftDestSpan.Length}");
+                    fixed (void* payload = &this)
+                    fixed (void* swiftDest = swiftDestSpan)
                     {
-                    metadata.ValueWitnessTable->InitializeWithCopy((void *)swiftDest, payload, metadata);
+                        metadata.ValueWitnessTable->InitializeWithCopy((void *)swiftDest, payload, metadata);
```

Comment:
The (void *) cast here looks unnecesary, `swiftDest` is already typed as `void*`:

```suggestion
                        metadata.ValueWitnessTable->InitializeWithCopy(swiftDest, payload, metadata);
```